### PR TITLE
[MM] Add Triton kernel for mm_input_normal.

### PR DESCRIPTION
**Pull request by noooop**
## Purpose

`mm_input_norm` is a memory-bandwidth-bound operator: one multiply-add per
element, so runtime is dominated by DRAM traffic. This PR adds a Triton
fusion kernel that performs the whole per-channel affine transform in a
single pass, replacing the eager path that materialized an fp32 intermediate.

The Triton fused kernel is **significantly faster** than the eager,
`BatchNorm1d`, and `torch.compile` baselines on the same workload, because it
reads the input once and writes the output once — no intermediate DRAM
round-trip.

https://github.com/noooop/snippet/tree/main/kernel_bench/mm_input_normal

image

This basically saturates my local 4090's VRAM bandwidth ～850GB/s.

image

```
./nvbandwidth -t device_local_copy_sm -b 1024

Running device_local_copy_sm.
memcpy SM local GPU(column) bandwidth (GB/s)
           0
 0   1695.46

SUM device_local_copy_sm 1695.46
 ```

1. **Single-pass fusion** — `y = x * weight[c] + bias[c]` computed in one
   kernel launch, at the input's native dtype (`uint8`/`fp16`/`bf16`/`fp32`).
2. **Preallocated output** — the caller can pass a larger buffer; only the
   `[:N]` region is written, so buffers can be reused across calls
   without a per-call allocation.
3. **`float32` compute for free** — arithmetic is done in fp32 inside the
   kernel regardless of I/O dtypes. This matches the eager path numerically
   but costs no bandwidth, since the intermediate never leaves registers.

## Test Plan

pytest tests/models/multimodal/generation_ppl_test/test_qwen.py
pytest tests/model_executor/layers/test_mm_input_norm.py

## Test Result

pass

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by cjackal** [comment]
While not yet mergeable per se, I think this direction sounds promising and (partly) resolves the scratchpad memory increase in #55370

File: vllm/model_executor/layers/fusion/mm_input_norm.py [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+
+import torch
+from torch import nn
+
+from vllm.config import ModelConfig
+from vllm.logger import init_logger
+from vllm.triton_utils import HAS_TRITON, triton
+
+if HAS_TRITON:
+    import triton.language as tl
```
Comment (Isotr0py) [inline vllm/model_executor/layers/fusion/mm_input_norm.py]:
```suggestion
from vllm.triton_utils import tl, triton
```
Use guarded triton/tl from `triton_utils` directly, no need to add `if HAS_TRITON:` statement.

File: vllm/model_executor/layers/fusion/mm_input_norm.py [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+            channel=channel,
+            dtype=torch.float32,
+        )
+
+    def forward(
+        self,
+        grid_thw: torch.Tensor,
+        visual_dtype: torch.dtype,
+        out: torch.Tensor | None = None,
+    ) -> torch.Tensor:
```
Comment (Isotr0py) [inline vllm/model_executor/layers/fusion/mm_input_norm.py]: Should we dispatch it through `CustomOp` instead?

File: vllm/model_executor/layers/fusion/mm_input_norm.py [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+        C: tl.constexpr,
+        HAS_MASK: tl.constexpr,
+        BLOCK_L: tl.constexpr,
+        COMPUTE_DTYPE: tl.constexpr,
+    ):
+        # 2D grid: (N, cdiv(L, BLOCK_L)). Each program processes all C channels
+        # for a single (n, L-block) tile, so the C loop is fully unrolled and
+        # no integer div/mod is needed to recover (n, c) from a flat pid.
+        n = tl.program_id(0)
+        lb = tl.program_id(1)
```
Comment (Isotr0py) [inline vllm/model_executor/layers/fusion/mm_input_norm.py]: Do we really need 2D grid layout just for an input norm kernel? I think fused input norm should be quite simple with 1D flat layout:
```python3
def _fused_input_norm_kernel(...):
    pid = tl.program_id(0)
    offs = pid.to(tl.int64) * BLOCK + tl.arange(0, BLOCK)
    mask = offs < numel
    c = (offs // L) % C
    x = tl.load(
        x_ptr + offs, mask=mask, other=0, eviction_policy="evict_first"
    ).to(COMPUTE_DTYPE)
    w = tl.load(w_ptr + c, mask=mask, other=0).to(COMPUTE_DTYPE)
    b = tl.load(b_ptr + c, mask=mask, other=0).to(COMPUTE_DTYPE)
    tl.store(
        y_ptr + offs, x * w + b, mask=mask, eviction_policy="evict_first"
    )
```

Comment (noooop) [reply vllm/model_executor/layers/fusion/mm_input_norm.py]: yes, we should only support outputs[:N] padding instead of outputs[:N, :C, :L].

**Comment by DarkLight1337** [comment]
Does this increase the startup time significantly? I feel that our startup time has degraded over the past few months because of the introduction of many triton kernels...

**Comment by noooop** [comment]
> Does this increase the startup time significantly? I feel that our startup time has degraded over the past few months because of the introduction of many triton kernels...

Because it is a very small kernel, it takes almost no time for startup.

tests/model_executor/layers/test_mm_input_norm.py

46 passed in 5.76s

***

I’m planning to also improve startup-time observability and find out what’s making startup slower and slower, if #56427 can move forward smoothly.

File: vllm/model_executor/layers/fusion/mm_input_norm.py [hunk 4]
```diff
@@ -0,0 +1,586 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/fusion/mm_input_norm.py:1]: Tbh I'm not sure why `fusion` directory really exists. You can see there are a bunch of fused ops in the parent directory `layers`

Comment (noooop) [reply vllm/model_executor/layers/fusion/mm_input_norm.py:1]: ┓( ´∀` )┏

***

We should move the fused ops in the parent directory layers to the fusion directory.

File: vllm/model_executor/layers/fusion/mm_input_norm.py (hunk 3 above)
Comment (noooop) [reply vllm/model_executor/layers/fusion/mm_input_norm.py]: image

The 1D flat layout is slightly faster.

File: vllm/model_executor/layers/fusion/mm_input_norm.py [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+    c = (offs // L) % C
+
+    x = tl.load(
+        x_ptr + offs,
+        mask=mask,
+        other=0,
+        eviction_policy="evict_first",
+    ).to(COMPUTE_DTYPE)
+    w = tl.load(w_ptr + c, mask=mask, other=0).to(COMPUTE_DTYPE)
+    b = tl.load(b_ptr + c, mask=mask, other=0).to(COMPUTE_DTYPE)
```
Comment (Isotr0py) [inline vllm/model_executor/layers/fusion/mm_input_norm.py]:
```suggestion
    x = tl.load(
        x_ptr + offs,
        mask=mask,
        other=0,
        eviction_policy="evict_first",
    ).to(tl.float32)
    w = tl.load(w_ptr + c, mask=mask, other=0).to(tl.float32)
    b = tl.load(b_ptr + c, mask=mask, other=0).to(tl.float32)
```
We can hardcode to fp32 compute dtype if we rely on cuda core ops.

File: vllm/model_executor/layers/fusion/mm_input_norm.py [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+    torch.float32: tl.float32,
+}
+
+_SUPPORTED_INPUTS = (
+    torch.uint8,
+    torch.float16,
+    torch.bfloat16,
+    torch.float32,
+)
+_SUPPORTED_OUTPUTS = (torch.float16, torch.bfloat16, torch.float32)
```
Comment (Isotr0py) [inline vllm/model_executor/layers/fusion/mm_input_norm.py]: We usually infer dtype from `x_ptr.dtype.element_ty` instead, so no need to add an extra `tl <-> torch` dtype mapping.

File: vllm/model_executor/layers/fusion/mm_input_norm.py [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+            return self.forward_native(grid_thw, visual_dtype, out)
+
+        assert size % self.channel == 0, (
+            f"size={size} is not divisible by channel={self.channel}"
+        )
+        patch_size = size // self.channel
+
+        # Materialize a contiguous copy once if needed; the kernel
+        # requires contiguous inputs.
+        x = grid_thw if grid_thw.is_contiguous() else grid_thw.contiguous()
```
Comment (Isotr0py) [inline vllm/model_executor/layers/fusion/mm_input_norm.py]:
```suggestion
        x = grid_thw.contiguous()
```

File: vllm/model_executor/layers/fusion/mm_input_norm.py [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+        self,
+        grid_thw: torch.Tensor,
+        visual_dtype: torch.dtype,
+        out: torch.Tensor | None = None,
+    ) -> torch.Tensor:
+        """Triton kernel path for CUDA devices."""
+        patches, size, out_view = self._prepare_output(grid_thw, visual_dtype, out)
+
+        if self.is_identity:
+            return self._identity_forward(grid_thw, visual_dtype, out_view)
```
Comment (Isotr0py) [inline vllm/model_executor/layers/fusion/mm_input_norm.py]: I think identity forwrad can be placed at `forward` before dispatching, so that we don't have to repeat it inside each platform's dispatch forward.

File: vllm/model_executor/layers/fusion/mm_input_norm.py [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+
+        # Fall back to native if the Triton kernel cannot service this call.
+        if (
+            not HAS_TRITON
+            or grid_thw.dtype not in _SUPPORTED_INPUTS
+            or visual_dtype not in _SUPPORTED_OUTPUTS
+            or self.weight.dtype != torch.float32
+            or not self.weight.is_contiguous()
+            or not self.bias.is_contiguous()
+        ):
```
Comment (Isotr0py) [inline vllm/model_executor/layers/fusion/mm_input_norm.py]: Is it a reasonable case that the cuda forward should fall back to native forwarding?

File: vllm/model_executor/layers/fusion/mm_input_norm.py [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+        # Normalize to concrete types so the return type is exactly as
+        # declared (``Any`` is not assignable to ``list[float]`` / ``float``
+        # without a cast or a value-level construction).
+        return (
+            bool(do_rescale),
+            bool(do_normalize),
+            [float(v) for v in image_mean],
+            [float(v) for v in image_std],
+            float(rescale_factor),
+        )
```
Comment (Isotr0py) [inline vllm/model_executor/layers/fusion/mm_input_norm.py]: Maybe use a namedtuple? It's terrible to identify a tuple with 5 elements. 🥵

File: vllm/model_executor/layers/fusion/mm_input_norm.py [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+    # 1D grid over the flattened (N, C, L) tensor. Each program processes
+    # BLOCK contiguous elements; the channel index is recovered from the
+    # flat offset via `(offs // L) % C`. int32 offsets: int64 division is
+    # ~2x slower here (measured on GB200); the wrapper asserts
+    # numel < 2**31 so the flat index never overflows.
+    pid = tl.program_id(0)
+    offs = pid * BLOCK + tl.arange(0, BLOCK)
+    mask = offs < numel
+    c = (offs // L) % C
+
```
Comment (noooop) [inline vllm/model_executor/layers/fusion/mm_input_norm.py]: int32 offsets +1 

I didn't observe this on the 4090, but on the H800 I found an 8% bandwidth difference compared with dummy torch.

**Review by Isotr0py — APPROVED** [review]
